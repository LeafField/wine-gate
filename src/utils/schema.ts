import { z } from "zod";

export const searchSchema = z.object({
  search: z
    .string()
    .min(4, { message: "4文字以上入力してください" })
    .max(60, { message: "60文字以下で入力してください" }),
  category: z.string(),
  sort: z.enum(["new", "inexpensive", "popular"]),
});
