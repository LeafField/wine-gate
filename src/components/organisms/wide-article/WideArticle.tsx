import React, { FC } from "react";
import { ArticleProps } from "../../../types/article_types";
import Link from "next/link";
import Image from "next/image";
import ArticleTags from "../../molecules/article-tags/ArticleTags";
import ArticleInfo from "../../molecules/article-info/ArticleInfo";

type Props = {
  article: ArticleProps;
  image_src: string;
};

const WideArticle: FC<Props> = ({ article, image_src }) => {
  return (
    <Link href={`/artcle/${article.id}`} className="block @container">
      <article className="group flex flex-col border border-gray transition-colors duration-300 hover:border-blue @[767px]:flex-row">
        <div className="aspect-video w-full shrink-0 overflow-hidden @[767px]:aspect-auto @[767px]:h-auto @[767px]:w-[20.8125rem]">
          <figure
            style={{ position: "relative" }}
            className="relative h-full w-full bg-gray transition-transform duration-300 group-hover:scale-110"
          >
            <Image
              src={image_src}
              alt={`${article.title}の画像`}
              fill
              style={{ objectFit: "contain", width: "100%", height: "100%" }}
              sizes="(min-width: 768px) 20.8125rem, 100vw"
              priority
            />
          </figure>
        </div>
        <div className="flex-1 space-y-4 px-3 py-5 ">
          <h3 className="text-heading3">{article.title}</h3>
          <ArticleTags tags={article.tags} category={article.categories} />
          <ArticleInfo
            red={article.categories.sub === "red"}
            sober_or_sweet={article.sober_or_sweet}
            tart={article.tart}
            fruity={article.fruity}
          />
          <div className="flex justify-between">
            <p className="text-main-text">{article.author_name}</p>
            <p className="text-main-text">{article.price}円</p>
          </div>
        </div>
      </article>
    </Link>
  );
};

export default WideArticle;
