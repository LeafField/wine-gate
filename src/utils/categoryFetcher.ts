import { supabase } from "./supabase";
import { cache } from "react";
import { ArticleProps } from "../types/article_types";

export const categoryFetcher = cache(
  async (slug: string[]): Promise<ArticleProps[] | null> => {
    if (slug.length !== 3) return null;
    const category = slug[0];
    const sort = slug[1];
    const page = Number(slug[2]);
    const start = (page - 1) * 10;
    const end = page * 10 - 1;
    if (sort === "new") {
      const { data, error } = await supabase
        .from("wines")
        .select(
          "title,sober_or_sweet,tart,fruity,author_name,tags,id,price,image_src,categories(*)",
        )
        .eq("categories.category", category)
        .order("created_at", { ascending: false })
        .range(start, end);
      if (error) throw new Error("カテゴリの取得に失敗しました");
      return data;
    } else if (sort === "inexpensive") {
      const { data, error } = await supabase
        .from("wines")
        .select(
          "title,sober_or_sweet,tart,fruity,author_name,tags,id,price,image_src,categories(*)",
        )
        .eq("categories.category", category)
        .order("price", { ascending: true })
        .range(start, end);
      if (error) throw new Error("カテゴリの取得に失敗しました");
      return data;
    } else if (sort === "popular") {
      const { data, error } = await supabase
        .rpc("get_wine_category_favorite", {
          category_value: category,
        })
        .limit(10);
      if (error) throw new Error("カテゴリの取得に失敗しました");
      const articles: ArticleProps[] = data.map((article) => ({
        categories: {
          category: article.category_name,
          id: article.category_id,
          chara: article.category_chara,
          sub: article.category_sub,
        },
        author_name: article.author_name,
        fruity: article.fruity,
        id: article.id,
        image_src: article.image_src,
        price: article.price,
        sober_or_sweet: article.sober_or_sweet,
        tags: article.tags,
        tart: article.tart,
        title: article.title,
      }));

      return articles;
    }
    return null;
  },
);

export const categoryCountFetcher = cache(async (category: string) => {
  const { count } = await supabase
    .from("wines")
    .select("id,categories(category)", { count: "exact", head: true })
    .eq("categories.category", category);
  if (count) {
    const totalPage = Math.ceil(count / 10);
    return totalPage;
  } else {
    return 1;
  }
});
