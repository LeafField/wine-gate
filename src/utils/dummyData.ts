import dummyImage from "../images/dummy_wine.png";
import {
  SmallArticleProps,
  CategoryProps,
  WineProps,
  ArticleProps,
} from "../types/article_types";

export const smallArticleDummyData: SmallArticleProps[] = [
  {
    id: "1",
    title: "コノスル カベルネソーヴィニヨン ビシクレタ・レゼルバ",
    chara: "渋い",
    tags: "樽熟成",
    image_src: dummyImage.src,
  },
  {
    id: "2",
    title: "コノスル シャルドネ ビシクレタ・レゼルバ",
    chara: "辛口白",
    tags: "フルーティ",
    image_src: dummyImage.src,
  },
  {
    id: "3",
    title: "コノスル カルメネール ビシクレタ・レゼルバ",
    chara: "酸っぱい",
    tags: "樽熟成,濃厚",
    image_src: dummyImage.src,
  },
];

export const categoryDummyData: CategoryProps[] = [
  {
    id: 1,
    category: "sober",
    chara: "渋い",
    sub: "red",
  },
  {
    id: 2,
    category: "tart",
    chara: "酸っぱい",
    sub: "red",
  },
  {
    id: 3,
    category: "fruity",
    chara: "果実味豊か",
    sub: "red",
  },
  {
    id: 4,
    category: "white",
    chara: "辛口白",
    sub: "white",
  },
  {
    id: 5,
    category: "sweet",
    chara: "甘口白",
    sub: "white",
  },
  {
    id: 6,
    category: "rose",
    chara: "ロゼ",
    sub: "interim",
  },
  {
    id: 7,
    category: "orange",
    chara: "オレンジ",
    sub: "interim",
  },
  {
    id: 8,
    category: "sparkling",
    chara: "スパークリング",
    sub: "other",
  },
  {
    id: 9,
    category: "fortified",
    chara: "酒精強化",
    sub: "other",
  },
];

export const selectDummyData = categoryDummyData.map((category) => {
  return { value: category.category, label: category.chara };
});

export const wineDummyData: WineProps = {
  id: "sdfaf",
  title: "コノスル カベルネソーヴィニヨン ビシクレタ・レゼルバ",
  author_id: "fsdaf",
  author_name: "ワイン大好き太郎",
  image_src: dummyImage.src,
  price: 2000,
  categories: categoryDummyData[0],
  tags: "樽熟成",
  sober_or_sweet: 3,
  tart: 3,
  fruity: 3,
  description:
    "チリのワインといえばコノスル！\nカベルネソーヴィニヨンの中でも特におすすめの一本です。",
  place: "スーパー,ネット通販",
  erudition:
    "カベルネソーヴィニヨンはボルドー地方の赤ワインの原料としても有名です。",
  category_id: 1,
  created_at: "2021-10-10",
};

export const articleDummyData: ArticleProps = {
  title: "コノスル カベルネソーヴィニヨン ビシクレタ・レゼルバ",
  author_name: "ワイン大好き太郎",
  tags: "樽熟成",
  sober_or_sweet: 3,
  tart: 3,
  fruity: 3,
  categories: categoryDummyData[0],
  id: "1",
  price: 1500,
};
