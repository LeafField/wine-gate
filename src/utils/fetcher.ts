import { supabase } from "./supabase";
import { EditingPageSchemaType, ImageInputSchemaType } from "./schema";
import { User } from "../store";
import { revalidateServer } from "../server/revalidate";

export const postWine = async ({
  value,
  user,
  image,
}: {
  value: EditingPageSchemaType;
  user: User;
  image: ImageInputSchemaType;
}) => {
  if (!image) throw new Error("画像が必須です");
  const IMAGE_SRC = `${user.id}/${image.name}`;

  const { error: imageError, data: imageData } = await supabase.storage
    .from("wines")
    .upload(IMAGE_SRC, image);

  if (imageError) throw new Error("画像のアップロードに失敗しました");
  if (imageData) {
    const { data, error } = await supabase
      .from("wines")
      .insert({
        title: value.title.replace(/　/g, " "),
        price: value.price,
        place: value.place,
        description: value.description,
        erudition: value.erudition,
        category_id: Number(value.category_id),
        fruity: value.fruity,
        tart: value.tart,
        sober_or_sweet: value.sober_or_sweet,
        tags: value.tags.join(),
        author_id: user.id,
        author_name: user.username,
        image_src: imageData.path,
      })
      .select();
    if (error) throw new Error("投稿に失敗しました");
    return data[0].id;
  }
};

export const updateWine = async ({
  id,
  image,
  oldWineImage,
  user_id,
  value,
}: {
  id: string;
  user_id: string;
  value: EditingPageSchemaType;
  image: ImageInputSchemaType | null;
  oldWineImage: string;
}) => {
  revalidateServer(id);
  if (image) {
    const IMAGE_SRC = `${user_id}/${image.name}`;
    const { error } = await supabase.storage
      .from("wines")
      .upload(IMAGE_SRC, image);
    if (error) throw new Error("画像のアップロードに失敗しました");
    const { error: deleteError } = await supabase.storage
      .from("wines")
      .remove([oldWineImage]);
    if (deleteError) throw new Error("画像の削除に失敗しました");
  }
  const { data, error } = await supabase
    .from("wines")
    .update({
      title: value.title.replace(/　/g, " "),
      price: value.price,
      place: value.place,
      description: value.description,
      erudition: value.erudition,
      category_id: Number(value.category_id),
      fruity: value.fruity,
      tart: value.tart,
      sober_or_sweet: value.sober_or_sweet,
      tags: value.tags.join(),
      image_src: image ? `${user_id}/${image.name}` : oldWineImage,
    })
    .eq("id", id)
    .select();
  if (error) throw new Error("ワインの更新に失敗しました");
  return data[0].id;
};

export const favoriteCount = async (wine_id: string) => {
  const { count, error } = await supabase
    .from("favorite")
    .select("id", { count: "exact", head: true })
    .eq("wine_id", wine_id);
  if (error) throw new Error("お気に入りのカウントに失敗しました");
  return count;
};

export const favoriteUserOne = async (wine_id: string) => {
  const { data, error } = await supabase
    .from("favorite")
    .select("user_id")
    .eq("wine_id", wine_id);
  if (error) throw new Error("お気に入りのユーザー取得に失敗しました");
  return data[0] ? data[0] : null;
};

export const addFavorite = async ({
  wine_id,
  user_id,
}: {
  wine_id: string;
  user_id: string;
}) => {
  const { error } = await supabase
    .from("favorite")
    .insert({ wine_id, user_id });

  if (error) throw new Error("お気に入りの登録に失敗しました");
};

export const removeFavorite = async ({
  wine_id,
  user_id,
}: {
  wine_id: string;
  user_id: string;
}) => {
  const { error } = await supabase
    .from("favorite")
    .delete()
    .eq("wine_id", wine_id)
    .eq("user_id", user_id);

  if (error) throw new Error("お気に入りの解除に失敗しました");
};

export const getWine = async (id: string) => {
  const { data, error } = await supabase
    .from("wines")
    .select("*,categories(*)")
    .eq("id", id)
    .single();
  if (error) throw new Error("ワインの取得に失敗しました");

  return data;
};
