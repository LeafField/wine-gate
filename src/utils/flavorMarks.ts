import { ReactNode } from "react";

type Marks = {
  value: number;
  label: ReactNode;
};

const sweet: Marks[] = [
  {
    value: 1,
    label: "辛口",
  },
  {
    value: 2,
    label: "中辛",
  },
  {
    value: 3,
    label: "中口",
  },
  {
    value: 4,
    label: "中甘",
  },
  {
    value: 5,
    label: "甘口",
  },
];

export const tart: Marks[] = [
  {
    value: 1,
    label: "ない",
  },
  {
    value: 2,
    label: "弱い",
  },
  {
    value: 3,
    label: "普通",
  },
  {
    value: 4,
    label: "強め",
  },
  {
    value: 5,
    label: "酸っぱい",
  },
];
