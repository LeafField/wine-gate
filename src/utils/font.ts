import { Noto_Sans_JP, Kumar_One } from "next/font/google";

export const notoStyle = Noto_Sans_JP({ subsets: ["latin"] });
export const KumarOne = Kumar_One({ subsets: ["latin"], weight: "400" });
