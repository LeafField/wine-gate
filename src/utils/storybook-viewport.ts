export const mobileViewport = {
  viewport: {
    defaultViewport: "iphone6",
  },
};
