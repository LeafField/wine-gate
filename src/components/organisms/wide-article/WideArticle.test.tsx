import { screen, render } from "@testing-library/react";
import WideArticle from "./WideArticle";
import { articleDummyData } from "../../../utils/dummyData";
import dummyImage from "../../../images/dummy_wine2.jpg";

describe("WideArticleの単体テスト", () => {
  test("WideArticleコンポーネントが存在することを確認する", () => {
    const { container } = render(<WideArticle article={articleDummyData} />);
    expect(screen.getByText(articleDummyData.title)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("favorite_countのpropsが無い場合、お気に入り数の表示が無い", () => {
    render(<WideArticle article={articleDummyData} />);
    expect(screen.queryByTestId("fa-star")).not.toBeInTheDocument();
  });

  test("categoriesがnullの場合何も返さない", () => {
    const failDummyData = {
      ...articleDummyData,
    };
    failDummyData.categories = null;

    render(<WideArticle article={failDummyData} />);
  });

  test("お気に入り数のデータがある場合、お気に入り数が表示される", () => {
    const favoriteDummyData = {
      ...articleDummyData,
      favorite: {
        count: 10,
      },
    };
    render(<WideArticle article={favoriteDummyData} />);
    expect(screen.getByTestId("fa-star")).toBeInTheDocument();
    expect(screen.getByText("10")).toBeInTheDocument();
  });
});
