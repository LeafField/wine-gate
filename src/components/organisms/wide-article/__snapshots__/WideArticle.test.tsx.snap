// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WideArticleの単体テスト WideArticleコンポーネントが存在することを確認する 1`] = `
<div>
  <a
    class="block @container"
    href="/article/4be1f2ab-b3b3-4599-b1a1-a73d280f5033"
  >
    <article
      class="group flex flex-col border border-gray transition-colors duration-300 hover:border-blue @[767px]:flex-row"
    >
      <div
        class="aspect-video w-full shrink-0 overflow-hidden @[767px]:aspect-auto @[767px]:h-auto @[767px]:w-[20.8125rem]"
      >
        <figure
          class="relative h-full w-full bg-gray transition-transform duration-300 group-hover:scale-110"
          style="position: relative;"
        >
          <img
            alt="コノスル カベルネソーヴィニヨン ビシクレタ・レゼルバの画像"
            data-nimg="fill"
            decoding="async"
            fetchpriority="high"
            loading="eager"
            sizes="(min-width: 768px) 20.8125rem, 100vw"
            src="/_next/image?url=%2Fimg.jpg&w=3840&q=75"
            srcset="/_next/image?url=%2Fimg.jpg&w=640&q=75 640w, /_next/image?url=%2Fimg.jpg&w=750&q=75 750w, /_next/image?url=%2Fimg.jpg&w=828&q=75 828w, /_next/image?url=%2Fimg.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fimg.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fimg.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fimg.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fimg.jpg&w=3840&q=75 3840w"
            style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; object-fit: contain; color: transparent;"
          />
        </figure>
      </div>
      <div
        class="flex-1 space-y-4 p-2 @[767px]:px-3 @[767px]:py-5 "
      >
        <h3
          class="text-heading3"
        >
          コノスル カベルネソーヴィニヨン ビシクレタ・レゼルバ
        </h3>
        <div
          class="flex w-fit items-center gap-2"
        >
          <div
            class="w-fit rounded-[3.2rem] border border-black px-2 py-[0.125rem] text-black text-base"
          >
            渋い
          </div>
          <div
            class="w-fit rounded-[3.2rem] border border-black px-2 py-[0.125rem] text-black text-base"
          >
            樽熟成
          </div>
        </div>
        <div
          class="flex w-fit gap-x-4"
        >
          <p
            class="text-main-text"
          >
            渋味
            :
            3
          </p>
          <p
            class="text-main-text"
          >
            酸味
            :
            3
          </p>
          <p
            class="text-main-text"
          >
            果実味
            :
            3
          </p>
          <div
            class="flex items-center"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-star text-sm text-yellow"
              data-icon="star"
              data-prefix="fas"
              data-testid="fa-star"
              focusable="false"
              role="img"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M316.9 18C311.6 7 300.4 0 288.1 0s-23.4 7-28.8 18L195 150.3 51.4 171.5c-12 1.8-22 10.2-25.7 21.7s-.7 24.2 7.9 32.7L137.8 329 113.2 474.7c-2 12 3 24.2 12.9 31.3s23 8 33.8 2.3l128.3-68.5 128.3 68.5c10.8 5.7 23.9 4.9 33.8-2.3s14.9-19.3 12.9-31.3L438.5 329 542.7 225.9c8.6-8.5 11.7-21.2 7.9-32.7s-13.7-19.9-25.7-21.7L381.2 150.3 316.9 18z"
                fill="currentColor"
              />
            </svg>
            :
            <span
              class="text-main-text"
            >
              10
            </span>
          </div>
        </div>
        <div
          class="flex justify-between"
        >
          <p
            class="text-main-text"
          >
            ワイン大好き太郎
          </p>
          <p
            class="text-main-text"
          >
            1500
            円
          </p>
        </div>
      </div>
    </article>
  </a>
</div>
`;
